app/code/Magento/InventoryCatalogAdminUi/Ui/DataProvider/Product/Listing/Modifier/QuantityPerSource.php

```diff
@@ -68,36 +87,58 @@ public function modifyData(array $data)
             return $data;
         }
 
-        foreach ($data['items'] as &$item) {
-            $item['quantity_per_source'] = $this->isSourceItemManagementAllowedForProductType->execute(
-                $item['type_id']
-            ) === true ? $this->getSourceItemsData($item['sku']) : [];
-        }
-        unset($item);
+        $data['items'] = $this->getSourceItemsData($data['items']);
 
         return $data;
     }
 
     /**
-     * @param string $sku
+     * @param array $dataItems
      * @return array
      * @throws NoSuchEntityException
      */
-    private function getSourceItemsData(string $sku): array
+    private function getSourceItemsData(array $dataItems): array
     {
-        $sourceItems = $this->getSourceItemsBySku->execute($sku);
+        $itemsBySkus = [];
+
+        foreach ($dataItems as $key => $item) {
+            if ($this->isSourceItemManagementAllowedForProductType->execute($item['type_id'])) {
```

Comment:
I do not think that it's efficient to call this service for each product in a loop, especially taking into account that there could be much of SourceItem data here and we have just limited number of Types.

You have this service 
\Magento\InventoryConfigurationApi\Model\GetAllowedProductTypesForSourceItemManagementInterface::execute
just call it once and provide product types from dataItems